gcamwrapper/query_library.py

```diff
@@ -0,0 +1,251 @@
+    '''Look up a query from a YAML file.
+       The YAML query files will often be organized into nested categories
+       so users must provide the full path as individual strings such as:
+       \code{get_query("emissions", "co2_emissions")}
+
+    :param query_file: The path to a YAML file to read if not None otherwise
+                       the package query file will be used.
+    :type query_file: str
+    :param *args: Strings specifiying the path to the query to lookup
+    :type *args: str
+
+    :returns: The raw query from the query file with units set as an attribute
+              if specified
+    '''
+
+    if query_file is None:
+        queries = PACKAGE_QUERIES
+    else:
+        queries = read_yaml(query_file)
+    query = queries
+    for path in args:
+        try:
+            query = query[path]
+        except KeyError:
+            raise Exception("Could not find query: ", args)
+
+    if len(query) == 0:
+        raise Exception("Could not find query: ", args)
+    query_str = Query(query[0])
+    if len(query) > 1:
+        query_str.units = query[1]
+
+    return query_str
+
+# Design doc:
+# change get/set data to accept a query_args argument that is a dict/list
+# used to substitute {arg_tag@arg_type} tags from the query database with the rules:
+# 1) arg_type = name -> NamedFilter; year -> YearFilter
+# 2) key: None -> +MatchesAny
+# 3) key: [operator, int-ish]
+#    where int-ish is ether an int or '.' support others?
+#    and operators are </>|= or * for any
+#    NOTE: we should support IndexFilter instead of YearFilter but to keep it simple
+#    for the users we will try to detect which they meant by assuming only years will
+#    be four digits long
+# 4) key: [ operator,char]
+#    where operators are either = or =~ for regex or * for any
+# 5) For query tags not specified by the user it will be replaced with nothing:
+# i.e. matches any but collapse
+
+def find_placeholders(query_str):
+    '''Find all placeholder in a raw query and return them in a list.
+       Find query placeholders of the form \code{'{tag@type}'} and put them into a
+       dict such as \code{dict('tag1': 'name', 'tag2': 'year')} for instance.
+
+    :param query_str: A raw query string
+    :type query_str: str
+
+    :returns: A dict of all the placeholders found in the query where the keys
+              are the tags and the values are the types
+    '''
+
+    raw_placeholders = re.findall(r'{([\w@]+)}', query_str)
+    ret = dict()
+    for placeholder in raw_placeholders:
+        ph_split = placeholder.split('@')
+        if len(ph_split) != 2:
+            raise Exception('Invalid placeholder syntax: ', placeholder, 'expecting two values split by @')
+        elif ph_split[1] != 'name' and ph_split[1] != 'year':
+            raise Exception('Invalid placeholder syntax, unknown type: ', ph_split[1], 'expecting name or year')
+        else:
+            ret[ph_split[0]] = ph_split[1]
```

Comment:
this doesn't catch duplicate values per key.  Is this case possible?  E.g., can `ph_split[0]` ever be the same?  If so, it could be overwritten.